Feature Request: add an additional argument to auto_optim to allow for gradient accumulation

When we use horovod backend and perform gradient accumulation, we get the following error:
`AssertionError: Gradients were computed more than backward_passes_per_step times before call to step(). Increase backward_passes_per_step to accumulate gradients
locally.`

Thus, we need to change the default argument `backward_passes_per_step` of `horovod.DistributedOptimizer` to enable gradient accumulation in the distributed setting.  To do so, we can add this argument to `ignite.distributed.auto_optim`.

--- a/ignite/distributed/auto.py
+++ b/ignite/distributed/auto.py
@@ -265,6 +265,9 @@ def auto_optim(optimizer: Optimizer, **kwargs: Any) -> Optimizer:
 
     .. versionchanged:: 0.4.2
         Added Horovod distributed optimizer.
+
+    .. versionchanged:: 0.5.0
+        Added kwargs to ``idist.auto_optim``.
     """
     bnd = idist.backend()
     if idist.has_xla_support and bnd == idist_xla.XLA_TPU:

--- a/ignite/distributed/auto.py
+++ b/ignite/distributed/auto.py
@@ -256,7 +256,7 @@ def auto_optim(optimizer: Optimizer,**kwargs: Any) -> Optimizer:
 
     Args:
         optimizer: input torch optimizer
-        kwargs: kwargs to horovod backend's DistributedOptimizer() method. 
+        kwargs: kwargs to horovod backend's DistributedOptimizer. 
 
     Returns:
         Optimizer

--- a/ignite/distributed/auto.py
+++ b/ignite/distributed/auto.py
@@ -234,7 +234,7 @@ def auto_model(model: nn.Module, sync_bn: bool = False, **kwargs: Any) -> nn.Mod
     return model
 
 
-def auto_optim(optimizer: Optimizer) -> Optimizer:
+def auto_optim(optimizer: Optimizer,**kwargs: Any) -> Optimizer:
     """Helper method to adapt optimizer for non-distributed and distributed configurations (supporting
     all available backends from :meth:`~ignite.distributed.utils.available_backends()`).
 
@@ -256,6 +256,7 @@ def auto_optim(optimizer: Optimizer) -> Optimizer:
 
     Args:
         optimizer: input torch optimizer
+        kwargs: kwargs to horovod backend's DistributedOptimizer() method. 
 
     Returns:
         Optimizer
@@ -273,7 +274,7 @@ def auto_optim(optimizer: Optimizer) -> Optimizer:
     if idist.has_hvd_support and bnd == idist_hvd.HOROVOD:
         import horovod.torch as hvd
 
-        optimizer = hvd.DistributedOptimizer(optimizer)
+        optimizer = hvd.DistributedOptimizer(optimizer,**kwargs)
         hvd.broadcast_optimizer_state(optimizer, root_rank=0)
         return optimizer
 

--- a/ignite/distributed/auto.py
+++ b/ignite/distributed/auto.py
@@ -234,7 +234,7 @@ def auto_model(model: nn.Module, sync_bn: bool = False, **kwargs: Any) -> nn.Mod
     return model
 
 
-def auto_optim(optimizer: Optimizer,**kwargs: Any) -> Optimizer:
+def auto_optim(optimizer: Optimizer, **kwargs: Any) -> Optimizer:
     """Helper method to adapt optimizer for non-distributed and distributed configurations (supporting
     all available backends from :meth:`~ignite.distributed.utils.available_backends()`).
 
@@ -256,7 +256,7 @@ def auto_optim(optimizer: Optimizer,**kwargs: Any) -> Optimizer:
 
     Args:
         optimizer: input torch optimizer
-        kwargs: kwargs to horovod backend's DistributedOptimizer. 
+        kwargs: kwargs to Horovod backend's DistributedOptimizer.
 
     Returns:
         Optimizer
@@ -274,7 +274,7 @@ def auto_optim(optimizer: Optimizer,**kwargs: Any) -> Optimizer:
     if idist.has_hvd_support and bnd == idist_hvd.HOROVOD:
         import horovod.torch as hvd
 
-        optimizer = hvd.DistributedOptimizer(optimizer,**kwargs)
+        optimizer = hvd.DistributedOptimizer(optimizer, **kwargs)
         hvd.broadcast_optimizer_state(optimizer, root_rank=0)
         return optimizer
 

--- a/tests/ignite/distributed/test_auto.py
+++ b/tests/ignite/distributed/test_auto.py
@@ -164,13 +164,19 @@ def _test_auto_model_optimizer(ws, device):
     # Test auto_optim
     bnd = idist.backend()
     optimizer = optim.SGD(model.parameters(), lr=0.01)
-    optimizer = auto_optim(optimizer, backward_passes_per_step=2)
+    optimizer = auto_optim(optimizer)
     if idist.has_xla_support and "xla" in device:
         assert isinstance(optimizer, optim.SGD) and hasattr(optimizer, "wrapped_optimizer")
     elif idist.has_hvd_support and bnd in ("horovod",):
         assert isinstance(optimizer, optim.SGD) and hasattr(optimizer, "_allreduce_grad_async")
     else:
         assert isinstance(optimizer, optim.SGD) and not hasattr(optimizer, "wrapped_optimizer")
+    if idist.has_hvd_support and bnd in ("horovod",):
+        backward_passes_per_step = 2
+        optimizer = optim.SGD(model.parameters(), lr=0.01)
+        optimizer = auto_optim(optimizer, backward_passes_per_step=backward_passes_per_step)
+        assert isinstance(optimizer, optim.SGD) and hasattr(optimizer, "backward_passes_per_step")
+        assert optimizer.backward_passes_per_step == backward_passes_per_step
 
 
 def test_auto_methods_no_dist():

--- a/tests/ignite/distributed/test_auto.py
+++ b/tests/ignite/distributed/test_auto.py
@@ -171,6 +171,7 @@ def _test_auto_model_optimizer(ws, device):
         assert isinstance(optimizer, optim.SGD) and hasattr(optimizer, "_allreduce_grad_async")
     else:
         assert isinstance(optimizer, optim.SGD) and not hasattr(optimizer, "wrapped_optimizer")
+
     if idist.has_hvd_support and bnd in ("horovod",):
         backward_passes_per_step = 2
         optimizer = optim.SGD(model.parameters(), lr=0.01)

--- a/tests/ignite/distributed/test_auto.py
+++ b/tests/ignite/distributed/test_auto.py
@@ -164,7 +164,7 @@ def _test_auto_model_optimizer(ws, device):
     # Test auto_optim
     bnd = idist.backend()
     optimizer = optim.SGD(model.parameters(), lr=0.01)
-    optimizer = auto_optim(optimizer, backward_passes_per_step = 2)
+    optimizer = auto_optim(optimizer, backward_passes_per_step=2)
     if idist.has_xla_support and "xla" in device:
         assert isinstance(optimizer, optim.SGD) and hasattr(optimizer, "wrapped_optimizer")
     elif idist.has_hvd_support and bnd in ("horovod",):

--- a/tests/ignite/distributed/test_auto.py
+++ b/tests/ignite/distributed/test_auto.py
@@ -164,7 +164,7 @@ def _test_auto_model_optimizer(ws, device):
     # Test auto_optim
     bnd = idist.backend()
     optimizer = optim.SGD(model.parameters(), lr=0.01)
-    optimizer = auto_optim(optimizer)
+    optimizer = auto_optim(optimizer, backward_passes_per_step = 2)
     if idist.has_xla_support and "xla" in device:
         assert isinstance(optimizer, optim.SGD) and hasattr(optimizer, "wrapped_optimizer")
     elif idist.has_hvd_support and bnd in ("horovod",):

